File: clients/transcript.ts
/**
 * Tiny client for the YouTube Transcript API.
 * Works in Deno, Node 18+, Bun, browsers, and Cloudflare Workers.
 */
export type TranscriptFormat = "json" | "plain" | "timestamps" | "paragraphs" | "srt" | "vtt";

export interface Segment {
  text: string;
  start: number;
  duration: number;
}

export interface TranscriptResponse {
  success: true;
  cached: boolean;
  video: { video_id: string; title: string; author: string; duration_seconds: number };
  format: TranscriptFormat;
  language: { code: string; name: string; kind: "asr" | "manual" };
  stats: { segment_count: number; word_count: number; character_count: number };
  available_languages: { language_code: string; name: string; kind: string }[];
  transcript: string | Segment[];
}

export class TranscriptApiError extends Error {
  constructor(readonly code: string, message: string, readonly status: number) {
    super(message);
    this.name = "TranscriptApiError";
  }
}

export interface ClientOptions {
  baseUrl: string;
  apiKey?: string;
}

export interface GetOptions {
  format?: TranscriptFormat;
  /** Priority-ordered language tags, e.g. ["hi", "en"]. */
  lang?: string[];
}

export function createTranscriptClient({ baseUrl, apiKey }: ClientOptions) {
  const origin = baseUrl.replace(/\/+$/, "");
  const headers: Record<string, string> = apiKey ? { "X-API-Key": apiKey } : {};

  async function request(path: string, params: Record<string, string>) {
    const url = `${origin}${path}?${new URLSearchParams(params)}`;
    const response = await fetch(url, { headers });
    const body = await response.json();
    if (!response.ok) {
      throw new TranscriptApiError(
        body?.error?.code ?? "unknown",
        body?.error?.message ?? response.statusText,
        response.status,
      );
    }
    return body;
  }

  return {
    /** Full JSON response including metadata. */
    async get(video: string, options: GetOptions = {}): Promise<TranscriptResponse> {
      return await request("/api/transcript", {
        url: video,
        format: options.format ?? "json",
        lang: (options.lang ?? ["en"]).join(","),
      });
    },

    /** Just the transcript text in the requested format. */
    async text(video: string, options: GetOptions = {}): Promise<string> {
      const result = await this.get(video, { format: options.format ?? "plain", ...options });
      return typeof result.transcript === "string"
        ? result.transcript
        : JSON.stringify(result.transcript);
    },

    /** Timed segments, useful for search or summarisation pipelines. */
    async segments(video: string, options: GetOptions = {}): Promise<Segment[]> {
      const result = await this.get(video, { ...options, format: "json" });
      return result.transcript as Segment[];
    },

    async languages(video: string) {
      return await request("/api/languages", { url: video });
    },
  };
}

File: clients/transcript.py
"""Tiny client for the YouTube Transcript API (requires `requests`)."""

from __future__ import annotations

from typing import Any, Iterable

import requests


class TranscriptApiError(RuntimeError):
    def __init__(self, code: str, message: str, status: int) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code
        self.status = status


class TranscriptClient:
    def __init__(self, base_url: str, api_key: str | None = None, timeout: int = 30) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        if api_key:
            self.session.headers["X-API-Key"] = api_key

    def _request(self, path: str, params: dict[str, str]) -> dict[str, Any]:
        response = self.session.get(f"{self.base_url}{path}", params=params, timeout=self.timeout)
        body = response.json()
        if not response.ok:
            error = body.get("error", {})
            raise TranscriptApiError(
                error.get("code", "unknown"),
                error.get("message", response.reason),
                response.status_code,
            )
        return body

    def get(
        self,
        video: str,
        fmt: str = "json",
        lang: Iterable[str] = ("en",),
    ) -> dict[str, Any]:
        """Full JSON response: metadata, stats, and transcript."""
        return self._request(
            "/api/transcript",
            {"url": video, "format": fmt, "lang": ",".join(lang)},
        )

    def text(self, video: str, fmt: str = "plain", lang: Iterable[str] = ("en",)) -> str:
        return self.get(video, fmt=fmt, lang=lang)["transcript"]

    def segments(self, video: str, lang: Iterable[str] = ("en",)) -> list[dict[str, Any]]:
        return self.get(video, fmt="json", lang=lang)["transcript"]

    def languages(self, video: str) -> dict[str, Any]:
        return self._request("/api/languages", {"url": video})


if __name__ == "__main__":
    import sys

    client = TranscriptClient(sys.argv[1])
    print(client.text(sys.argv[2]))
